### PR TITLE
Add support for ErrorPath, CausePath, MaxDelaySeconds, and JitterStrategy

### PR DESCRIPTION
**Pull request by evscott**
Adding support for ErrorPath, CausePath, MaxDelaySeconds, and JitterStrategy.

https://aws.amazon.com/about-aws/whats-new/2023/09/aws-step-functions-enhanced-error-handling/

https://docs.aws.amazon.com/step-functions/latest/dg/concepts-error-handling.html

### CODE REVIEW

File: data/StateMachine.j2119 [hunk 1]
```diff
@@ -28,7 +28,9 @@ A Pass State MAY have a field named "Result".
 A Fail State MUST NOT have a field named "InputPath".
 A Fail State MUST NOT have a field named "OutputPath".
 A Fail State MAY have a string field named "Cause".
+A Fail State MAY have a referencePath field named "CausePath".
```
Comment (yaythomas) [inline data/StateMachine.j2119:31]: `referencePath` only covers _some_ of the functionality.

These fields _also_ allow Intrinsic Functions that return a string. [Here's the full list](https://github.com/aws/amazon-states-language-service/blob/84829de062bfc4c0bbb5f644ae49a4f2c3b8338a/src/validation/validateStates.ts#L33)

As is, the referencePath rule will fail valid state machines, for example:
```
{
  "StartAt": "FailState",
  "States": {
    "FailState": {
      "Type": "Fail",
      "ErrorPath": "States.Format('this is the error: {}, and this is the cause: {}', $.Error, $.Cause)",
      "CausePath": "States.UUID()"
    }
  }
}
```

**Review by yaythomas — CHANGES_REQUESTED** [review]

File: data/StateMachine.j2119 (hunk 1 above)
Comment (evscott) [reply data/StateMachine.j2119:31]: True, we'll have to just accept any field as is done for `ResultSelector`.

File: data/StateMachine.j2119 [hunk 2]
```diff
@@ -28,7 +28,9 @@ A Pass State MAY have a field named "Result".
...
 A Fail State MAY have a string field named "Error".
+A Fail State MAY have a referencePath field named "ErrorPath".
```
[the remark this answers is not in the document]
Comment (evscott) [reply data/StateMachine.j2119:33]: Ack.

File: data/StateMachine.j2119 [hunk 3]
```diff
@@ -41,6 +43,8 @@ A Retrier MUST have a nonempty-string-array field named "ErrorEquals".
 A Retrier MAY have an positive-integer field named "IntervalSeconds".
 A Retrier MAY have a nonnegative-integer field named "MaxAttempts" whose value MUST be less than 99999999.
 A Retrier MAY have a float field named "BackoffRate" whose value MUST be greater than or equal to 1.0.
+A Retrier MAY have an positive-integer field named "MaxDelaySeconds".
```
Comment (abdelnn) [inline data/StateMachine.j2119:46]: small typo: a* vs. an (seems like its spread out across the file)

File: test/fail-with-error-and-cause-path.json
Comment (abdelnn) [inline test/fail-with-error-and-cause-path.json:1]: Could we also add tests for mutual exclusivity of `Error/Cause` and their `*Path` equivalent?

File: data/StateMachine.j2119
Comment (abdelnn) [inline data/StateMachine.j2119:1]: Could you write down in the description how those changes were tested?

File: test/fail-with-error-and-cause-path.json
Comment (evscott) [reply test/fail-with-error-and-cause-path.json:1]: Yeah, adding.

File: data/StateMachine.j2119
Comment (evscott) [reply data/StateMachine.j2119:1]: By running `rspec`.

File: data/StateMachine.j2119 (hunk 3 above)
Comment (evscott) [reply data/StateMachine.j2119:46]: Fixing.

**Review by abdelnn — APPROVED** [review]

**Review by yaythomas — APPROVED** [review]
Tightening up the validation of the new Path fields to JsonPath and Intrinsic Functions can be a separate PR. Currently these fields accept any given value, even non-strings.

Still TODO is to increment the version in gemspec in root and publish update.

**Comment by itstheceo** [comment]
Any ETA on when these changes will be released? Thanks :)

**Comment by yaythomas** [comment]
PR is up now, please see #54

**Pull request merged**